File: apps/web/src/lib/api/catalog.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { z } from "zod";

import type { WorkspaceContext } from "./client";
import {
  createOpenApiClient,
  parseOpenApiResult,
  requestJsonWithFetch,
  workspaceHeaders,
  workspaceQueryParams,
} from "./core";
import { buildApiHeaders } from "./client";
import { queryKeys } from "./query-keys";
import { healthBandSchema, nullableStringSchema, supportTierSchema } from "./schema-helpers";

const workspaceContextCompanyProfileSchema = z.object({
  id: z.string(),
  legal_name: z.string(),
  sector: nullableStringSchema,
  headquarters: nullableStringSchema,
  description: nullableStringSchema,
  ceo_name: nullableStringSchema,
  ceo_message: nullableStringSchema,
  sustainability_approach: nullableStringSchema,
  is_configured: z.boolean(),
});

const workspaceContextBrandKitSchema = z.object({
  id: z.string(),
  brand_name: z.string(),
  logo_uri: nullableStringSchema,
  primary_color: z.string(),
  secondary_color: z.string(),
  accent_color: z.string(),
  font_family_headings: z.string(),
  font_family_body: z.string(),
  tone_name: nullableStringSchema,
  is_configured: z.boolean(),
});

export const workspaceIntegrationSummarySchema = z.object({
  id: z.string(),
  connector_type: z.string(),
  display_name: z.string(),
  status: z.string(),
  support_tier: supportTierSchema,
  certified_variant: nullableStringSchema,
  product_version: nullableStringSchema,
  health_band: healthBandSchema,
  assigned_agent_status: nullableStringSchema,
  last_discovered_at: nullableStringSchema,
  last_preflight_at: nullableStringSchema,
  last_preview_sync_at: nullableStringSchema,
  last_synced_at: nullableStringSchema,
});

export const factoryReadinessSchema = z.object({
  is_ready: z.boolean(),
  company_profile_ready: z.boolean(),
  brand_kit_ready: z.boolean(),
  blockers: z.array(
    z.object({
      code: z.string(),
      message: z.string(),
    }),
  ),
});

export const workspaceContextResponseSchema = z.object({
  tenant: z.object({
    id: z.string(),
    name: z.string(),
    slug: z.string(),
    status: z.string(),
  }),
  project: z.object({
    id: z.string(),
    tenant_id: z.string(),
    name: z.string(),
    code: z.string(),
    reporting_currency: z.string(),
    status: z.string(),
  }),
  company_profile: workspaceContextCompanyProfileSchema,
  brand_kit: workspaceContextBrandKitSchema,
  integrations: z.array(workspaceIntegrationSummarySchema),
  blueprint_version: z.string(),
  factory_readiness: factoryReadinessSchema,
});

export const workspaceBootstrapResponseSchema = workspaceContextResponseSchema.extend({
  tenant_created: z.boolean(),
  project_created: z.boolean(),
});

export const brandKitLogoUploadResponseSchema = z.object({
  logo_uri: z.string().trim().min(1),
  filename: z.string().trim().min(1),
  content_type: z.string().trim().min(1),
  size_bytes: z.number().int().nonnegative(),
});

export const workspaceBootstrapRequestSchema = z.object({
  tenantHeader: z.string().trim().min(1),
  tenant_name: z.string().trim().min(2),
  tenant_slug: z.string().trim().min(2),
  project_name: z.string().trim().min(2),
  project_code: z.string().trim().min(2),
  reporting_currency: z.string().trim().min(1),
  company_profile: z.object({
    legal_name: z.string().trim().min(1),
    sector: z.string().trim(),
    headquarters: z.string().trim(),
    description: z.string().trim(),
    ceo_name: z.string().trim(),
    ceo_message: z.string().trim(),
    sustainability_approach: z.string().trim(),
  }),
  brand_kit: z.object({
    brand_name: z.string().trim().min(1),
    logo_uri: z.string().trim(),
    primary_color: z.string().trim().min(1),
    secondary_color: z.string().trim().min(1),
    accent_color: z.string().trim().min(1),
    font_family_headings: z.string().trim().min(1),
    font_family_body: z.string().trim().min(1),
    tone_name: z.string().trim(),
  }),
});

export const integrationSyncRequestSchema = z.object({
  tenant_id: z.string().trim().min(1),
  project_id: z.string().trim().min(1),
  connector_ids: z.array(z.string().trim().min(1)).min(1),
});

export const integrationSyncResponseSchema = z.object({}).passthrough();
export const brandKitLogoUploadRequestSchema = z.object({
  tenantHeader: z.string().trim().min(1),
  tenantId: nullableStringSchema.optional(),
  projectId: nullableStringSchema.optional(),
});

export type WorkspaceContextResponse = z.infer<typeof workspaceContextResponseSchema>;
export type WorkspaceBootstrapResponse = z.infer<typeof workspaceBootstrapResponseSchema>;
export type WorkspaceIntegrationSummary = z.infer<typeof workspaceIntegrationSummarySchema>;
export type WorkspaceBootstrapRequest = z.infer<typeof workspaceBootstrapRequestSchema>;
export type IntegrationSyncRequest = z.infer<typeof integrationSyncRequestSchema>;
export type BrandKitLogoUploadResponse = z.infer<typeof brandKitLogoUploadResponseSchema>;
export type BrandKitLogoUploadRequest = z.infer<typeof brandKitLogoUploadRequestSchema> & {
  file: File;
};

export async function fetchWorkspaceContext(
  workspace: WorkspaceContext,
  signal?: AbortSignal,
): Promise<WorkspaceContextResponse> {
  const client = createOpenApiClient();
  return parseOpenApiResult(
    await client.GET("/catalog/workspace-context", {
      params: {
        query: workspaceQueryParams(workspace),
      },
      headers: workspaceHeaders({ workspace }),
      signal,
    }),
    workspaceContextResponseSchema,
  );
}

export async function bootstrapWorkspace(
  input: WorkspaceBootstrapRequest,
): Promise<WorkspaceBootstrapResponse> {
  const parsed = workspaceBootstrapRequestSchema.parse(input);
  const client = createOpenApiClient();

  return parseOpenApiResult(
    await client.POST("/catalog/bootstrap-workspace", {
      headers: {
        "x-tenant-id": parsed.tenantHeader,
        "x-user-id": "web-ui-user",
        "x-user-role": "analyst",
        "Content-Type": "application/json",
      },
      body: {
        tenant_name: parsed.tenant_name,
        tenant_slug: parsed.tenant_slug,
        project_name: parsed.project_name,
        project_code: parsed.project_code,
        reporting_currency: parsed.reporting_currency,
        company_profile: parsed.company_profile,
        brand_kit: parsed.brand_kit,
      },
    }),
    workspaceBootstrapResponseSchema,
  );
}

export async function uploadBrandKitLogo(
  input: BrandKitLogoUploadRequest,
): Promise<BrandKitLogoUploadResponse> {
  if (!(input.file instanceof File)) {
    throw new Error("Select a valid logo file before uploading.");
  }

  const parsed = brandKitLogoUploadRequestSchema.parse(input);
  const formData = new FormData();
  formData.append("file", input.file);
  if (parsed.tenantId) {
    formData.append("tenant_id", parsed.tenantId);
  }
  if (parsed.projectId) {
    formData.append("project_id", parsed.projectId);
  }

  return requestJsonWithFetch(
    "/catalog/brand-kit-logo",
    {
      method: "POST",
      headers: buildApiHeaders(parsed.tenantHeader, {
        includeJsonContentType: false,
      }),
      body: formData,
    },
    brandKitLogoUploadResponseSchema,
  );
}

export async function syncIntegrations(request: IntegrationSyncRequest, signal?: AbortSignal) {
  const parsed = integrationSyncRequestSchema.parse(request);
  const client = createOpenApiClient();

  return parseOpenApiResult(
    await client.POST("/integrations/sync", {
      headers: {
        "x-tenant-id": parsed.tenant_id,
        "x-user-id": "web-ui-user",
        "x-user-role": "analyst",
        "Content-Type": "application/json",
      },
      body: parsed,
      signal,
    }),
    integrationSyncResponseSchema,
  );
}

export function useWorkspaceContextQuery(workspace: WorkspaceContext | null) {
  return useQuery({
    queryKey: queryKeys.catalog.workspaceContext(workspace),
    queryFn: ({ signal }) => fetchWorkspaceContext(workspace!, signal),
    enabled: Boolean(workspace),
  });
}

export function useBootstrapWorkspaceMutation() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: bootstrapWorkspace,
    onSuccess: async () => {
      await Promise.all([
        queryClient.invalidateQueries({ queryKey: ["catalog"] }),
        queryClient.invalidateQueries({ queryKey: ["dashboard"] }),
        queryClient.invalidateQueries({ queryKey: ["runs"] }),
        queryClient.invalidateQueries({ queryKey: ["integrations"] }),
      ]);
    },
  });
}

File: apps/api/app/schemas/catalog.py
# Bu sema dosyasi, catalog icin API veri sozlesmelerini tanimlar.

from __future__ import annotations

from pydantic import BaseModel, Field


class TenantCreateRequest(BaseModel):
    name: str = Field(min_length=2, max_length=200)
    slug: str = Field(min_length=2, max_length=120)


class TenantResponse(BaseModel):
    id: str
    name: str
    slug: str
    status: str


class ProjectCreateRequest(BaseModel):
    tenant_id: str = Field(min_length=1)
    name: str = Field(min_length=2, max_length=200)
    code: str = Field(min_length=2, max_length=64)
    reporting_currency: str = Field(default="TRY", min_length=1, max_length=8)


class ProjectResponse(BaseModel):
    id: str
    tenant_id: str
    name: str
    code: str
    reporting_currency: str
    status: str


class CompanyProfileResponse(BaseModel):
    id: str
    tenant_id: str
    project_id: str
    legal_name: str
    sector: str | None = None
    headquarters: str | None = None
    description: str | None = None
    ceo_name: str | None = None
    ceo_message: str | None = None
    sustainability_approach: str | None = None
    is_configured: bool = False


class BrandKitResponse(BaseModel):
    id: str
    tenant_id: str
    project_id: str
    brand_name: str
    logo_uri: str | None = None
    primary_color: str
    secondary_color: str
    accent_color: str
    font_family_headings: str
    font_family_body: str
    tone_name: str | None = None
    is_configured: bool = False


class BrandKitLogoUploadResponse(BaseModel):
    logo_uri: str
    filename: str
    content_type: str
    size_bytes: int


class FactoryReadinessBlockerResponse(BaseModel):
    code: str
    message: str


class FactoryReadinessResponse(BaseModel):
    is_ready: bool
    company_profile_ready: bool
    brand_kit_ready: bool
    blockers: list[FactoryReadinessBlockerResponse] = Field(default_factory=list)


class IntegrationConfigSummaryResponse(BaseModel):
    id: str
    connector_type: str
    display_name: str
    status: str
    support_tier: str
    certified_variant: str | None = None
    product_version: str | None = None
    health_band: str
    last_discovered_at: str | None = None
    last_preflight_at: str | None = None
    last_preview_sync_at: str | None = None
    last_synced_at: str | None = None
    assigned_agent_status: str | None = None


class TenantListResponse(BaseModel):
    items: list[TenantResponse]
    total: int


class ProjectListResponse(BaseModel):
    items: list[ProjectResponse]
    total: int


class WorkspaceBootstrapRequest(BaseModel):
    tenant_name: str = Field(min_length=2, max_length=200)
    tenant_slug: str = Field(min_length=2, max_length=120)
    project_name: str = Field(min_length=2, max_length=200)
    project_code: str = Field(min_length=2, max_length=64)
    reporting_currency: str = Field(default="TRY", min_length=1, max_length=8)
    company_profile: dict[str, str] = Field(default_factory=dict)
    brand_kit: dict[str, str] = Field(default_factory=dict)


class WorkspaceContextResponse(BaseModel):
    tenant: TenantResponse
    project: ProjectResponse
    company_profile: CompanyProfileResponse
    brand_kit: BrandKitResponse
    integrations: list[IntegrationConfigSummaryResponse]
    blueprint_version: str
    factory_readiness: FactoryReadinessResponse


class WorkspaceBootstrapResponse(WorkspaceContextResponse):
    tenant_created: bool
    project_created: bool
